add bulk submit commands to cheat sheet

vsoch opened the pull request:
Ping @grondo and @chu11 for review - thanks for the help and pointers today!

I really like bulk submit, especially the custom format strings and using my own inputs! And the dry run! Yeah this is totally something people will want to know about.

vsoch committed:
add bulk submit commands to cheat sheet

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

chu11 commented on _data/flux-commands.yaml:
```diff
@@ -119,6 +121,41 @@ commands:
+        items:
+          - title: Pipe inputs into flux bulksubmit
+            command: seq 1 4 | flux bulksubmit echo {}
+          - title: Use GNU parallel to map the same inputs in
```
It sounds like you're saying we're using gnu parallel, which I don't think we are.  I think you mean "gnu parallel style input" or something to that effect.

vsoch replied:
Oh I thoughts that is what the three dots were - @grondo what are the dots from?

grondo replied:
Do you mean the triple-colon syntax `:::`? That is borrowed from GNU Parallel, but we don't use or depend on GNU parallel in any way.

vsoch replied:
ahh gotcha, missed that part! I'll remove it.

chu11 commented on _data/flux-commands.yaml:
```diff
@@ -119,6 +121,41 @@ commands:
+          - title: Do a dry run and provide multiple inputs using GNU parallel
+            command: "flux bulksubmit --dry-run /bin/bash -c 'touch mydir/{0}; echo touched number {1}' ::: filea fileb filec filed ::: 1 2 3 4"
+          - title: Use flux 'carbon copy' to submit identical jobs with different inputs
+            command: flux submit --cc="1-10" echo "Hello I am job {cc}"
```
Idea: perhaps instead just show the bulksubmit example and add a fyi, could also be done with flux submit --cc like this ...

Also, could add some --cc examples to flux submit, and do a similar "you could also use bulksubmit like this ..."

vsoch replied:
I think if people are coming to the cheat sheet looking for "how do I submit my jobs in bulk" it would be ok to have the `--cc` example here!

But we definitely could add more examples up there - I only saw the one example from your docs, but if you want to do another follow up PR to add more extensive ones that would be great!

chu11 replied:
> I think if people are coming to the cheat sheet looking for "how do I submit my jobs in bulk" it would be ok to have the --cc example here!

I see your point.  I guess `bulksubmit` is simply the name we use for the gnu parallel style input.  Internally the guts are identical to `--cc` with submit.  edit: i.e. maybe it should have been `flux submit-parallel` or something, but that's sorta long.

grondo replied:
Hm, I wouldn't say the internals are the same as `--cc`, in fact, with `bulksubmit` you can substitute into the `--cc` argument:
```
$ flux bulksubmit --dry-run --cc={1} echo {} ::: a b c ::: 0-1 0-3 0-7
bulksubmit: submit --cc=0-1 echo a
bulksubmit: submit --cc=0-3 echo a
bulksubmit: submit --cc=0-7 echo a
bulksubmit: submit --cc=0-1 echo b
bulksubmit: submit --cc=0-3 echo b
bulksubmit: submit --cc=0-7 echo b
bulksubmit: submit --cc=0-1 echo c
bulksubmit: submit --cc=0-3 echo c
bulksubmit: submit --cc=0-7 echo c
```

The reason we didn't call it `flux parallel` or something like that is because "parallel" is overloaded in the context of Flux, since you can use `submit` to launch tasks in parallel, where this tool submits jobs in parallel. In that case `bulksubmit` seemed more appropriate and less confusing.

Edit: Though reading back what you wrote above @chu11, I think I see now what you meant and agree  - `bulksubmit` is really doing the same thing as many `flux submit` calls in parallel.

vsoch commented:
okay updated!

- I removed references to GNU parallel
- To go with @chu11 idea, I put an example for using carbon copy both under submit and bulksubmit. Redundancy doesn't hurt, especially when it's paired with two commands!
- I also split the (getting very long) `flux submit` top level section into two groups for basic and advanced.

grondo commented on _data/flux-commands.yaml:
```diff
@@ -119,7 +127,42 @@ commands:
+            command: "flux bulksubmit --dry-run /bin/bash -c 'touch mydir/{0}; echo touched number {1}' ::: filea fileb filec filed ::: 1 2 3 4"
+          - title: Use carbon copy to submit identical jobs with different inputs
+            command: "flux bulksubmit --dry-run --cc={0} echo {1} ::: a b c ::: 0-1 0-3 0-7"
+          - title: Similar functionality with flux bulksubmit
+            command: "flux bulksubmit echo 'Hello I am job {}' ::: $(seq 1 10)"
```
This one seems to be out of place. It says "similar functionality with `flux bulksubmit`", but the previous example also uses `bulksubmit`

vsoch replied:
Good catch - this ordering has changed (so the context is outdated) - will try to write a better description!

grondo commented on _data/flux-commands.yaml:
```diff
@@ -119,7 +127,42 @@ commands:
+        items:
+          - title: Pipe inputs into flux bulksubmit
+            command: seq 1 4 | flux bulksubmit echo {}
+          - title: Map the same inputs into flux bulksubmit
```
Maybe something like "Map the same inputs on the command line using GNU parallel-style `:::` separator" (give credit where credit is due)

grondo commented on _data/flux-commands.yaml:
```diff
@@ -119,7 +127,42 @@ commands:
+          - title: Compress every file under ./mydir in parallel, one process per core
+            command: find ./mydir | flux bulksubmit --wait xz {} 
+          - title: Do a dry run and provide multiple inputs
+            command: "flux bulksubmit --dry-run /bin/bash -c 'touch mydir/{0}; echo touched number {1}' ::: filea fileb filec filed ::: 1 2 3 4"
```
I'm not sure this example works the way you might think it does. Just to double check, here's the result:
```console
$ flux bulksubmit --dry-run /bin/bash -c 'touch mydir/{0}; echo touched number {1}' ::: filea fileb filec filed ::: 1 2 3 4
bulksubmit: submit /bin/bash -c touch mydir/filea; echo touched number 1
bulksubmit: submit /bin/bash -c touch mydir/filea; echo touched number 2
bulksubmit: submit /bin/bash -c touch mydir/filea; echo touched number 3
bulksubmit: submit /bin/bash -c touch mydir/filea; echo touched number 4
bulksubmit: submit /bin/bash -c touch mydir/fileb; echo touched number 1
bulksubmit: submit /bin/bash -c touch mydir/fileb; echo touched number 2
bulksubmit: submit /bin/bash -c touch mydir/fileb; echo touched number 3
bulksubmit: submit /bin/bash -c touch mydir/fileb; echo touched number 4
bulksubmit: submit /bin/bash -c touch mydir/filec; echo touched number 1
bulksubmit: submit /bin/bash -c touch mydir/filec; echo touched number 2
bulksubmit: submit /bin/bash -c touch mydir/filec; echo touched number 3
bulksubmit: submit /bin/bash -c touch mydir/filec; echo touched number 4
bulksubmit: submit /bin/bash -c touch mydir/filed; echo touched number 1
bulksubmit: submit /bin/bash -c touch mydir/filed; echo touched number 2
bulksubmit: submit /bin/bash -c touch mydir/filed; echo touched number 3
bulksubmit: submit /bin/bash -c touch mydir/filed; echo touched number 4
```
Note that multiple inputs are combined, so in this example the same file gets touched 4 times, but with a different message for each based on the 2nd input. Maybe what you wanted was something that used the combinations like:
```console
$ flux bulksubmit --dry-run /bin/bash -c 'touch mydir/{0}/{1}; echo touched {0}/{1}' ::: dira dirb dirc dird ::: {1..4}
bulksubmit: submit /bin/bash -c touch mydir/dira/1; echo touched dira/1
bulksubmit: submit /bin/bash -c touch mydir/dira/2; echo touched dira/2
bulksubmit: submit /bin/bash -c touch mydir/dira/3; echo touched dira/3
bulksubmit: submit /bin/bash -c touch mydir/dira/4; echo touched dira/4
bulksubmit: submit /bin/bash -c touch mydir/dirb/1; echo touched dirb/1
bulksubmit: submit /bin/bash -c touch mydir/dirb/2; echo touched dirb/2
bulksubmit: submit /bin/bash -c touch mydir/dirb/3; echo touched dirb/3
bulksubmit: submit /bin/bash -c touch mydir/dirb/4; echo touched dirb/4
bulksubmit: submit /bin/bash -c touch mydir/dirc/1; echo touched dirc/1
bulksubmit: submit /bin/bash -c touch mydir/dirc/2; echo touched dirc/2
bulksubmit: submit /bin/bash -c touch mydir/dirc/3; echo touched dirc/3
bulksubmit: submit /bin/bash -c touch mydir/dirc/4; echo touched dirc/4
bulksubmit: submit /bin/bash -c touch mydir/dird/1; echo touched dird/1
bulksubmit: submit /bin/bash -c touch mydir/dird/2; echo touched dird/2
bulksubmit: submit /bin/bash -c touch mydir/dird/3; echo touched dird/3
bulksubmit: submit /bin/bash -c touch mydir/dird/4; echo touched dird/4
```

vsoch replied:
Oh agree this is better - I think your original example had a submit with the parameter, and me changing that to a touch (so I could test and see the result) was an example that didn't make a ton of sense. I think I'll update the above to be mkdir -p too since the directory is unlikely to exist.

vsoch commented:
Updated 
![image](https://user-images.githubusercontent.com/814322/233406346-4f817e5d-767f-4783-9fdb-915dce9b4c9b.png)

chu11 commented on _data/flux-commands.yaml:
```diff
@@ -13,13 +13,19 @@ commands:
+
+      - name: Advanced job submission
+        items:
+          - title: Use flux carbon copy to submit identical jobs with different inputs
+            command: flux submit --cc="1-10" echo "Hello I am job {cc}"
```
Would it be worthwhile to add an example like:

```
# submits scripts myscript1.sh through myscript10.sh  
flux submit --cc="1-10" myscript{cc}.sh
```

(maybe w/ bulksubmit too)

just not sure if average user can see leap from the echo examples to this?  Or maybe I don't give average user enough credit.

grondo replied:
Good idea IMO. Note: the argument to --cc doesn't need to be quoted.

vsoch replied:
okay added! Good to go now?

vsoch commented:
okay I think this is good to go - going to merge. Thanks @grondo and @chu11 !
